### PR TITLE
 Laboratory work #4, Artyom Krivtsov - 22FPL2

### PR DESCRIPTION
**Pull request by artyom6k**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -12,7 +12,6 @@ class WordProcessor(TextProcessor):
     """
     Handle text tokenization, encoding and decoding.
     """
-
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:15]: revert

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,7 +27,17 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
-
+        if not (isinstance(text, str) and text):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:31]: add message to the error

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -28,7 +27,17 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+        clean_txt = []
+        for word in text.lower().split():
+            if word[-1] in '!?.':
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:34]: `!?.` to a variable

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -28,7 +27,17 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
...
+                clean_txt.extend([word[:-1], self._end_of_word_token])
+            else:
+                clean_word = [char for char in word if char.isalpha()]
+                if clean_word:
+                    clean_txt.append(''.join(clean_word))
+        return tuple(clean_txt)
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:40]: add empty line

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
What about start.py?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Where is the start.py?

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (artyom6k) [reply lab_4_fill_words_by_ngrams/main.py:40]: Done

File: lab_4_fill_words_by_ngrams/main.py (hunk 2 above)
Comment (artyom6k) [reply lab_4_fill_words_by_ngrams/main.py:31]: Done

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (artyom6k) [reply lab_4_fill_words_by_ngrams/main.py:15]: Done

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**